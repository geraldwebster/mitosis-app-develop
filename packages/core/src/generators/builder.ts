import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';

export type ToBuilderOptions = {
  prettier?: boolean;
};

export const blockToBuilder = (json: JSXLiteNode, options: ToBuilderOptions = {}) => {
  return null;
};

export const componentToBuilder = (
  componentJson: JSXLiteComponent,
  options: ToBuilderOptions = {},
) => {
  // Make a copy we can safely mutate, similar to babel's toolchain
  return null;
};
