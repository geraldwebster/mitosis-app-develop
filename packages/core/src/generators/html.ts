import { camelCase } from 'lodash';
import { format } from 'prettier/standalone';
import { getStateObjectString } from '../helpers/get-state-object-string';
import { collectCss } from '../helpers/collect-styles';
import { fastClone } from '../helpers/fast-clone';
import { selfClosingTags } from '../parsers/jsx';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';
import { dashCase } from '../helpers/dash-case';
import { isComponent } from '../helpers/is-component';
import { methodLiteralPrefix } from '../constants/method-literal-prefix';
import traverse from 'traverse';
import { isJsxLiteNode } from '../helpers/is-jsx-lite-node';
import { replaceIdentifiers } from '../helpers/replace-idenifiers';

type ToHtmlOptions = {
  prettier?: boolean;
};

type StringMap = { [key: string]: string };
type InternalToHtmlOptions = ToHtmlOptions & {
  onChangeJsById: StringMap;
  js: string;
};

const getGetters = (json: JSXLiteComponent) => {
  const getters: string[] = [];
  for (const key in json.state) {
    const value = json.state[key];
    if (typeof value === 'string') {
      if (value.startsWith(methodLiteralPrefix + 'get ')) {
        getters.push(key);
      }
    }
  }
  return getters;
};

const getForNames = (json: JSXLiteComponent) => {
  const names: string[] = [];
  traverse(json).forEach(function (item) {
    if (isJsxLiteNode(item)) {
      if (item.name === 'For') {
        names.push(item.bindings._forName as string);
      }
    }
  });
  return names;
};

const replaceForNameIdentifiers = (json: JSXLiteComponent) => {
  // TODO: cache this. by reference? lru?
  const forNames = getForNames(json);

  traverse(json).forEach((item) => {
    if (isJsxLiteNode(item)) {
      for (const key in item.bindings) {
        if (key === 'css' || key === '_forName') {
          continue;
        }
        const value = item.bindings[key];
        if (typeof value === 'string') {
          item.bindings[key] = replaceIdentifiers(
            value,
            forNames,
            (name) => `getContext(el, "${name}")`,
          ) as string;
        }
      }
    }
  });
};

const mappers: {
  [key: string]: (json: JSXLiteNode, options: InternalToHtmlOptions) => string;
} = {
  Fragment: (json, options) => {
    return json.children.map((item) => blockToHtml(item, options)).join('\n');
  },
};

const getId = (json: JSXLiteNode) =>
  `${dashCase(json.name)}-${Math.random().toString(26).slice(9)}`;

const addOnChangeJs = (
  id: string,
  options: InternalToHtmlOptions,
  code: string,
) => {
  if (!options.onChangeJsById[id]) {
    options.onChangeJsById[id] = '';
  }
  options.onChangeJsById[id] += code;
};

// TODO: spread support
const blockToHtml = (json: JSXLiteNode, options: InternalToHtmlOptions) => {
  const hasData = Object.keys(json.bindings).length;
  let elId = '';
  if (hasData) {
    elId = getId(json);
    json.properties['data-uid'] = elId;
  }

  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  if (json.properties._text) {
    return json.properties._text;
  }
  if (json.bindings._text) {
    addOnChangeJs(elId, options, `el.innerText = ${json.bindings._text};`);

    return `<span data-uid="${elId}"></span>`;
  }

  let str = '';

  if (json.name === 'For') {
    const itemName = json.bindings._forName;
    addOnChangeJs(
      elId,
      options,
      `
        el.innerHTML = '';
        var _array = ${json.bindings.each};
        var template = document.querySelector('[data-template="${elId}"]');
        for (var value of _array) {
          var tmp = document.createElement('span');
          tmp.innerHTML = template.innerHTML;
          Array.from(tmp.children).forEach(function (child) {
            contextMap.set(child, {
              ...contextMap.get(child),
              ${itemName}: value
            });
            el.appendChild(child);
          });
        }
      `,
    );
    // TODO: decide on how to handle this...
    str += `
      <span data-uid="${elId}"></span>
      <template data-template="${elId}">`;
    if (json.children) {
      str += json.children.map((item) => blockToHtml(item, options)).join('\n');
    }
    str += '</template>';
  } else if (json.name === 'Show') {
    addOnChangeJs(
      elId,
      options,
      `el.style.display = ${json.bindings.when} ? 'inline' : 'none'`,
    );

    str += `<span data-uid="${elId}">`;
    if (json.children) {
      str += json.children.map((item) => blockToHtml(item, options)).join('\n');
    }

    str += '</span>';
  } else {
    str += `<${json.name} `;

    // For now, spread is not supported
    // if (json.bindings._spread === '_spread') {
    //   str += `
    //       {% for _attr in ${json.bindings._spread} %}
    //         {{ _attr[0] }}="{{ _attr[1] }}"
    //       {% endfor %}
    //     `;
    // }

    for (const key in json.properties) {
      const value = json.properties[key];
      str += ` ${key}="${value}" `;
    }

    for (const key in json.bindings) {
      if (key === '_spread' || key === 'ref' || key === 'css') {
        continue;
      }
      const value = json.bindings[key] as string;
      // TODO: proper babel transform to replace. Util for this
      const useValue = value;

      if (key.startsWith('on')) {
        let event = key.replace('on', '').toLowerCase();
        if (!isComponent(json) && event === 'change') {
          event = 'input';
        }
        const fnName = camelCase(`on-${elId}-${event}`);
        options.js += `
          function ${fnName}(event) {
            ${useValue}
          }
        `;
        addOnChangeJs(
          elId,
          options,
          `
            el.removeEventListener('${event}', ${fnName});
            el.addEventListener('${event}', ${fnName});
          `,
        );
      } else {
        const useAttribute = key.includes('-');
        addOnChangeJs(
          elId,
          options,
          useAttribute
            ? `el.setAttribute(${key}, ${useValue})`
            : `el.${key} = ${useValue}`,
        );
      }
    }
    if (selfClosingTags.has(json.name)) {
      return str + ' />';
    }
    str += '>';
    if (json.children) {
      str += json.children.map((item) => blockToHtml(item, options)).join('\n');
    }

    str += `</${json.name}>`;
  }
  return str;
};

// TODO: props support via custom elements
export const componentToHtml = (
  componentJson: JSXLiteComponent,
  options: ToHtmlOptions = {},
) => {
  const useOptions: InternalToHtmlOptions = {
    ...options,
    onChangeJsById: {},
    js: '',
  };
  const json = fastClone(componentJson);
  replaceForNameIdentifiers(json);

  const hasState = Boolean(Object.keys(json.state).length);

  const css = collectCss(json);
  let str = json.children
    .map((item) => blockToHtml(item, useOptions))
    .join('\n');

  if (css.trim().length) {
    str += `<style>${css}</style>`;
  }

  if (hasState) {
    // TODO: collectJs helper for here and liquid
    str += `
      <script>
        var rawState = ${getStateObjectString(json)};

        var observers = [];
        var contextMap = new WeakMap();
        var state = new Proxy(rawState, {
          set(target, key, value, receiver) {
            Reflect.set(target, key, value, receiver);
            runObservers(key, value, receiver)
          },
        });

        onChange(function () {
          ${Object.keys(useOptions.onChangeJsById)
            .map((key) => {
              const value = useOptions.onChangeJsById[key];
              if (!value) {
                return '';
              }
              return `
              document.querySelectorAll("[data-uid='${key}']").forEach(function (el) {
                ${value}
              })
            `;
            })
            .join('\n\n')}
        })
        runObservers()

        ${useOptions.js}

        function runObservers(key, value, receiver) {
          observers.forEach(function (cb) {
            cb(key, value, receiver);
          })
        }
        function onChange(cb) {
          observers.push(cb);
        }
        function getContext(el, name) {
          var parent = el;
          do {
            var context = contextMap.get(parent);
            if (context && name in context) {
              return context[name];
            }
          } while (parent = parent.parentNode)
        }
      </script>
    `;
  }

  if (options.prettier !== false) {
    try {
      str = format(str, {
        parser: 'html',
        plugins: [
          // To support running in browsers
          require('prettier/parser-html'),
          require('prettier/parser-postcss'),
          require('prettier/parser-babel'),
        ],
      });
    } catch (err) {
      console.warn('Could not prettify', { string: str }, err);
    }
  }
  return str;
};
