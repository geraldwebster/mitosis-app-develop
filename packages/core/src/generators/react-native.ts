import json5 from 'json5';
import { camelCase, size } from 'lodash';
import { fastClone } from '../helpers/fast-clone';
import traverse from 'traverse';
import { ClassStyleMap } from '../helpers/collect-styles';
import { isJsxLiteNode } from '../helpers/is-jsx-lite-node';
import { Plugin } from '../modules/plugins';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { componentToReact } from './react';

type ToReactNativeOptions = {
  prettier?: boolean;
  stylesType?: 'emotion' | 'react-native';
  stateType?: 'useState' | 'mobx' | 'valtio' | 'solid' | 'builder';
  plugins?: Plugin[];
};

export const collectReactNativeStyles = (
  json: JSXLiteComponent,
): ClassStyleMap => {
  const styleMap: ClassStyleMap = {};

  const componentIndexes: { [className: string]: number | undefined } = {};

  traverse(json).forEach(function (item) {
    if (isJsxLiteNode(item)) {
      if (typeof item.bindings.css === 'string') {
        const value = json5.parse(item.bindings.css);
        delete item.bindings.css;
        if (!size(value)) {
          return;
        }
        const componentName = camelCase(item.name || 'view');
        const index = (componentIndexes[componentName] =
          (componentIndexes[componentName] || 0) + 1);
        const className = `${componentName}${index}`;
        item.bindings.style = `styles.${className}`;

        styleMap[className] = value;
      }
    }
  });

  return styleMap;
};

export const componentToReactNative = (
  componentJson: JSXLiteComponent,
  options: ToReactNativeOptions,
) => {
  const json = fastClone(componentJson);
  traverse(json).forEach((node) => {
    if (isJsxLiteNode(node)) {
      // TODO: handle TextInput, Image, etc
      if (node.name.toLowerCase() === node.name) {
        node.name = 'View';
      }

      if (
        node.properties._text?.trim().length ||
        node.bindings._text?.trim()?.length
      ) {
        node.name = 'Text';
      }
    }
  });

  return componentToReact(json, {
    ...options,
    stylesType: options.stylesType || 'react-native',
    type: 'native',
  });
};
