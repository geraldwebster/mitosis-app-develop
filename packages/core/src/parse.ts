import * as babel from '@babel/core';
import generate from '@babel/generator';
import JSON5 from 'json5';
import { JSXLiteNode } from './types/jsx-lite-node';
import { JSONObject, JSONOrNode, JSONOrNodeObject } from './types/json';
import { createJSXLiteNode } from './helpers/create-jsx-lite-node';
import { JSXLiteComponent, JSXLiteImport } from './types/jsx-lite-component';
import { createJSXLiteComponent } from './helpers/create-jsx-lite-component';

const jsxPlugin = require('@babel/plugin-syntax-jsx');
const tsPreset = require('@babel/preset-typescript');

export const selfClosingTags = new Set(['input', 'meta', 'bar']);

const { types } = babel;

type Context = {
  // Babel has other context
  builder: {
    component: JSXLiteComponent;
  };
};

const arrayToAst = (array: JSONOrNode[]) => {
  return types.arrayExpression(array.map((item) => jsonToAst(item)) as any);
};

const jsonToAst = (json: JSONOrNode): babel.Node => {
  if (types.isNode(json as any)) {
    if (types.isJSXText(json as any)) {
      return types.stringLiteral((json as any).value);
    }
    return json as babel.Node;
  }
  switch (typeof json) {
    case 'undefined':
      return types.identifier('undefined');
    case 'string':
      return types.stringLiteral(json);
    case 'number':
      return types.numericLiteral(json);
    case 'boolean':
      return types.booleanLiteral(json);
    case 'object':
      if (!json) {
        return types.nullLiteral();
      }
      if (Array.isArray(json)) {
        return arrayToAst(json);
      }
      return jsonObjectToAst(json as JSONObject);
  }
};

const jsonObjectToAst = (
  json: JSONOrNodeObject,
): babel.types.ObjectExpression => {
  if (!json) {
    return json;
  }
  const properties: babel.types.ObjectProperty[] = [];
  for (const key in json) {
    const value = json[key];
    if (value === undefined) {
      continue;
    }
    const keyAst = types.stringLiteral(key);
    const valueAst = jsonToAst(value);
    properties.push(types.objectProperty(keyAst, valueAst as any));
  }
  const newNode = types.objectExpression(properties);

  return newNode;
};

const componentFunctionToJson = (
  node: babel.types.FunctionDeclaration,
  context: Context,
): JSONOrNode => {
  let state = {};
  for (const item of node.body.body) {
    if (types.isVariableDeclaration(item)) {
      const init = item.declarations[0].init;
      if (types.isCallExpression(init)) {
        const firstArg = init.arguments[0];
        if (types.isObjectExpression(firstArg)) {
          const obj = JSON5.parse(generate(firstArg).code);
          state = obj;
        }
      }
    }
  }

  const theReturn = node.body.body.find((item) =>
    types.isReturnStatement(item),
  );
  const children: JSXLiteNode[] = [];
  if (theReturn) {
    const value = (theReturn as babel.types.ReturnStatement).argument;
    if (types.isJSXElement(value)) {
      children.push(jsxElementToJson(value) as JSXLiteNode);
    }
  }

  return {
    ...context.builder.component,
    state,
    children,
  } as any;
};

const jsxElementToJson = (
  node: babel.types.JSXElement | babel.types.JSXText,
): JSXLiteNode => {
  if (types.isJSXText(node)) {
    return createJSXLiteNode({
      properties: {
        _text: node.value,
      },
    });
  }
  return createJSXLiteNode({
    name: (node.openingElement.name as babel.types.JSXIdentifier).name,
    properties: node.openingElement.attributes.reduce((memo, item) => {
      if (types.isJSXAttribute(item)) {
        const key = item.name.name as string;
        const value = item.value;
        if (types.isStringLiteral(value)) {
          memo[key] = value;
          return memo;
        }
      }
      return memo;
    }, {} as { [key: string]: JSONOrNode }) as any,
    bindings: node.openingElement.attributes.reduce((memo, item) => {
      if (types.isJSXAttribute(item)) {
        const key = item.name.name as string;
        const value = item.value;

        if (types.isJSXExpressionContainer(value)) {
          const { expression } = value;
          if (types.isArrowFunctionExpression(expression)) {
            memo[key] = generate(expression.body).code;
          } else {
            memo[key] = generate(expression).code;
          }
          return memo;
        }
      } else if (types.isJSXSpreadAttribute(item)) {
        // TODO: potentially like Vue store bindings and properties as array of key value pairs
        // too so can do this accurately when order matters. Also tempting to not support spread,
        // as some frameworks do not support it (e.g. Angular) tho Angular may be the only one
        memo._spread = types.stringLiteral(generate(item.argument).code);
      }
      return memo;
    }, {} as { [key: string]: JSONOrNode }) as any,
    children: node.children.map((item) => jsxElementToJson(item as any)) as any,
  });
};

export function parse(jsx: string): JSXLiteComponent {
  const output = babel.transform(jsx, {
    presets: [[tsPreset, { isTSX: true, allExtensions: true }]],
    plugins: [
      jsxPlugin,
      () => ({
        visitor: {
          Program(path: babel.NodePath<babel.types.Program>, context: Context) {
            if (context.builder) {
              return;
            }
            context.builder = {
              component: createJSXLiteComponent(),
            };

            const isImportOrDefaultExport = (node: babel.Node) =>
              types.isExportDefaultDeclaration(node) ||
              types.isImportDeclaration(node);

            const keepStatements = path.node.body.filter((statement) =>
              isImportOrDefaultExport(statement),
            );
            const cutStatements = path.node.body.filter(
              (statement) => !isImportOrDefaultExport(statement),
            );

            // TODO: support multiple? e.g. for others to add imports?
            context.builder.component.hooks.preComponent = generate(
              types.program(cutStatements),
            ).code;

            path.replaceWith(types.program(keepStatements));
          },
          FunctionDeclaration(
            path: babel.NodePath<babel.types.FunctionDeclaration>,
            context: Context,
          ) {
            const { node } = path;
            if (types.isIdentifier(node.id)) {
              const name = node.id.name;
              if (name[0].toUpperCase() === name[0]) {
                path.replaceWith(
                  jsonToAst(componentFunctionToJson(node, context)),
                );
              }
            }
          },
          ImportDeclaration(
            path: babel.NodePath<babel.types.ImportDeclaration>,
            context: Context,
          ) {
            const importObject: JSXLiteImport = {
              imports: {},
              path: path.node.source.value,
            };
            for (const specifier of path.node.specifiers) {
              if (types.isImportSpecifier(specifier)) {
                importObject.imports[
                  (specifier.imported as babel.types.Identifier).name
                ] = specifier.local.name;
              } else if (types.isImportDefaultSpecifier(specifier)) {
                importObject.imports[specifier.local.name] = 'default';
              } else if (types.isImportNamespaceSpecifier(specifier)) {
                importObject.imports[specifier.local.name] = '*';
              }
            }
            context.builder.component.imports.push(importObject);
            // console.log('context 1', context);
            path.remove();
          },
          ExportDefaultDeclaration(
            path: babel.NodePath<babel.types.ExportDefaultDeclaration>,
          ) {
            path.replaceWith(path.node.declaration);
          },
          JSXText(path: babel.NodePath<babel.types.JSXText>) {
            path.replaceWith(types.stringLiteral(path.node.value));
          },
          JSXElement(path: babel.NodePath<babel.types.JSXElement>) {
            const { node } = path;
            path.replaceWith(jsonToAst(jsxElementToJson(node)));
          },
        },
      }),
    ],
  });

  try {
    return JSON5.parse(
      output!.code!.trim().replace(/^\({/, '{').replace(/}\);$/, '}'),
    );
  } catch (err) {
    console.error('Could not parse code', output && output.code);
    throw err;
  }
}
