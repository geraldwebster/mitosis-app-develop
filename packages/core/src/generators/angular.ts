import dedent from 'dedent';
import { format } from 'prettier';
import { renderPreComponent } from '../helpers/render-imports';
import { selfClosingTags } from '../parse';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';

type ToAngularOptions = {
  prettier?: boolean;
};
const blockToAngular = (json: JSXLiteNode, options: ToAngularOptions = {}) => {
  if (json.properties._text) {
    return json.properties._text;
  }

  let str = `<${json.name} `;
  for (const key in json.properties) {
    const value = json.properties[key];
    str += ` ${key}="${value}" `;
  }
  for (const key in json.bindings) {
    const value = json.bindings[key] as string;
    // TODO: proper babel transform to replace. Util for this
    const useValue = value.replace(/state\./g, '');

    if (key.startsWith('on')) {
      const event = key.replace('on', '').toLowerCase();
      str += ` (${event})="${useValue}" `;
    } else {
      str += ` [${key}]="${useValue}" `;
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }
  str += '>';
  if (json.children) {
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
  }

  str += `</${json.name}>`;
  return str;
};

export const componentToAngular = (
  json: JSXLiteComponent,
  options: ToAngularOptions = {},
) => {
  let str = dedent`
    import { Component } from '@angular/core';
    ${renderPreComponent(json)}

    @Component({
      template: \`
        ${json.children.map((item) => blockToAngular(item)).join('\n')}
      \`
    })
    export default class MyComponent {
      ${Object.keys(json.state)
        .map((key) => ` ${key} = ${json.state[key]};`)
        .join('\n')}
    }
  `;

  if (options.prettier !== false) {
    str = format(str, { parser: 'typescript' });
  }
  return str;
};
