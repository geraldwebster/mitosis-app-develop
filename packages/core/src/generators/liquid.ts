import { format } from 'prettier';
import { selfClosingTags } from '../parse';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';

type ToLiquidOptions = {
  prettier?: boolean;
};
// TODO: spread support
const blockToLiquid = (json: JSXLiteNode, options: ToLiquidOptions = {}) => {
  if (json.properties._text) {
    return json.properties._text;
  }

  let str = `<${json.name} `;

  for (const key in json.properties) {
    const value = json.properties[key];
    str += ` ${key}="${value}" `;
  }

  for (const key in json.bindings) {
    const value = json.bindings[key] as string;
    // TODO: proper babel transform to replace. Util for this
    const useValue = value.replace(/state\./g, '');

    if (key.startsWith('on')) {
      // Do nothing
    } else {
      str += ` ${key}="{{${useValue}}}" `;
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }
  str += '>';
  if (json.children) {
    str += json.children.map((item) => blockToLiquid(item, options)).join('\n');
  }

  str += `</${json.name}>`;
  return str;
};

export const componentToLiquid = (
  json: JSXLiteComponent,
  options: ToLiquidOptions = {},
) => {
  let str = json.children.map((item) => blockToLiquid(item)).join('\n');

  if (options.prettier !== false) {
    str = format(str, { parser: 'html' });
  }
  return str;
};
