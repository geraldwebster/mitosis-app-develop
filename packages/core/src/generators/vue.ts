import dedent from 'dedent';
import json5 from 'json5';
import { format } from 'prettier';
import { renderImports } from '../helpers/render-imports';
import { selfClosingTags } from '../parse';
import { JSXLiteComponent } from '../types/jsx-lite-component';
import { JSXLiteNode } from '../types/jsx-lite-node';

export type ToVueOptions = {
  prettier?: boolean;
};
export const blockToVue = (json: JSXLiteNode, options: ToVueOptions = {}) => {
  if (json.properties._text) {
    return json.properties._text;
  }

  let str = `<${json.name} `;
  for (const key in json.properties) {
    const value = json.properties[key];
    str += ` ${key}="${value}" `;
  }
  for (const key in json.bindings) {
    const value = json.bindings[key] as string;
    // TODO: proper babel transform to replace. Util for this
    const useValue = value.replace(/state\./g, '');

    if (key.startsWith('on')) {
      const event = key.replace('on', '').toLowerCase();
      str += ` @${event}="${useValue}" `;
    } else {
      str += ` :${key}="${useValue}" `;
    }
  }
  if (selfClosingTags.has(json.name)) {
    return str + ' />';
  }
  str += '>';
  if (json.children) {
    str += json.children.map((item) => blockToVue(item, options)).join('\n');
  }

  str += `</${json.name}>`;
  return str;
};
export const componentToVue = (
  json: JSXLiteComponent,
  options: ToVueOptions = {},
) => {
  let str = dedent`
    <template>
      ${json.children.map((item) => blockToVue(item)).join('\n')}
    </template>
    <script>
      ${renderImports(json.imports)}
      
      export default {
        data: () => (${json5.stringify(json.state)})
      }
    </script>
  `;

  if (options.prettier !== false) {
    str = format(str, { parser: 'html' });
  }
  return str;
};
