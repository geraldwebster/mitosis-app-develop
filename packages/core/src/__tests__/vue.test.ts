import { componentToVue } from '../generators/vue';
import { parse } from '../parse';
const basic = require('./data/basic.raw');
const inputBlock = require('./data/blocks/input.raw');
const selectBlock = require('./data/blocks/select.raw');
const formBlock = require('./data/blocks/form.raw');
const submitButtonBlock = require('./data/blocks/submit-button.raw');

describe('Vue', () => {
  test('Basic', () => {
    const json = parse(basic);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Input block', () => {
    const json = parse(inputBlock);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Submit button block', () => {
    const json = parse(submitButtonBlock);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Select block', () => {
    const json = parse(selectBlock);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });

  test('Form block', () => {
    const json = parse(formBlock);
    const output = componentToVue(json);
    expect(output).toMatchSnapshot();
  });
});
