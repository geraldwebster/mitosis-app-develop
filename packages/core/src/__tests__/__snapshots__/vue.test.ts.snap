// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vue Basic 1`] = `
"<template>
  <div>
    <input :value=\\"name\\" @change=\\"name = $event.target.value\\" />

    Hello! I can run in React, Vue, Solid, or Liquid!
  </div>
</template>
<script>
  export default {
    data: () => ({ name: \\"Steve\\" }),
  };
</script>
"
`;

exports[`Vue Input block 1`] = `
"<template>
  <input
    v-bind=\\"props.attributes\\"
    :key=\\"Builder.isEditing && defaultValue ? defaultValue : 'default-key'\\"
    :placeholder=\\"placeholder\\"
    :type=\\"type\\"
    :name=\\"name\\"
    :value=\\"value\\"
    :defaultValue=\\"defaultValue\\"
    :required=\\"required\\"
  />
</template>
<script>
  import { Builder } from \\"@builder.io/sdk\\";

  export interface FormInputProps {
    type?: string;
    attributes?: any;
    name?: string;
    value?: string;
    placeholder?: string;
    defaultValue?: string;
    required?: boolean;
  }

  export default {
    data: () => ({ Builder }),
  };
</script>
"
`;

exports[`Vue Submit button block 1`] = `
"<template>
  <button v-bind=\\"this.props.attributes\\" type=\\"submit\\">{props.text}</button>
</template>
<script>
  export interface ButtonProps {
    attributes?: any;
    text?: string;
  }

  export default {
    data: () => ({}),
  };
</script>
"
`;
